### DISCUSSION TITLE
How do you set the slack variable matrices for dense QCQP framework?

### DISCUSSION THREAD

**Discussion by dlcfort** [giaf/hpipm#155] [Q&A] [answered]
How do you set the number of slack variables, and the slack variable matrices for the dense qcqp framework? I could not find functions to set Jsb, Jsg, or Jsq in "hpipm_d_dense_qcqp.h". Is the use of the slack variables not fully implemented for dense QCQP?

Also, is "ns" the total number of slack variables (number of lower and upper)? Or is it half the amount?

**Comment by giaf** [comment]
You can set the number of softed constraints using this and following routines https://github.com/giaf/hpipm/blob/master/include/hpipm_d_dense_qcqp_dim.h#L82 
I added the routines to set Jsb Jsg Jsq with this commit https://github.com/giaf/hpipm/commit/a7b7caa46568c4c42742f0dab70b4996118967a9

`ns` is the number of softed constraints, the actual number of slack variables depends on the type of constraint that is softed (but you don't need to set this in the solver, just `ns`).
Box and general affine constraints are two-sided, so there are two slacks per softed constraint (one for the upper bound and one for the lower bound).
Quadratic constraints are one-sided, so there is one slack per softed constraint.

**Reply by dlcfort** [reply]
Thank you for the quick update! So does ns = nsb + nsg + nsq? 

Also does nsb == nb, nsg == ng, and nsq == nq? If not how do you specify which linear inequality constraints are softed and which are not, or which quadratic inequality constraints are softed?  Do you just set the respective rows of Js for that slack variable equal to zero?

Are the number of columns for each matrix Jsb, Jsg, Jsq equal to ns?

**Comment by giaf** [comment, marked as answer]
Yes ns = nsb + nsg + nsq https://github.com/giaf/hpipm/blob/master/dense_qp/x_dense_qcqp_dim.c#L223

No, it holds 0<=nsb<=nb, 0<=nsg<=ng, 0<=nsq<=nq, as any number of constraints between 0 and all can be softed.
You can specify which component is softed in two ways:
1) by either their index using set_idxs (where constraints are ordered as bounds  first followed by general affine and finally quadratic, and with the indexing being 0-based).
2) by passing the J matrix, defined as the matrix specifying which constraint is softed by which slack, as e.g. the J^{s..} matrices in the QP formulations in https://arxiv.org/abs/2003.02547
So notice that J is assumed to be made of columns of the identity matrix and that columns should not be repeated (i.e. a single slack can not be used to soft multiple constraints)

**Reply by dlcfort** [reply]
Thank you this is all very clear! I implemented an optimization with slack variables however I am getting unexpected behavior when I try to use the L2 penalties. I set both Zl and Zu to identity matrices and no matter what weight I multiply Zl and Zu by it has no affect on the solution to the optimization.  However when I set the values of Zl and Zu to be negative, then it starts to drive the solution below the constraints.

The optimization works just fine with the L1 penalties (zu and zl), but the Zl and Zu matrices do not produce the behavior that is expected.

**Reply by giaf** [reply]
Not sure if this is the issue in your case, but Zl and Zu are assumed to be arrays of size `ns` that only contain the diagonal of the matrix (so not the full `ns x ns` matrix),  as these matrices can only be diagonal anyway in order to efficiently process them.

**Reply by dlcfort** [reply]
Changing that fixed the problem thank you!
